### DISCUSSION TITLE
Supporting jobs with large capacities

### DISCUSSION THREAD

**Discussion by Plastonick** [reinterpretcat/vrp#98] [Ideas]
Consider the problem of a fleet of vehicles with a range of different capacities.

If we have jobs whose "demand" is greater than the capacity of any vehicle, then that job clearly cannot be serviced by any single vehicle, but could possible be serviced by multiple trips of the same, or different vehicles.

Is there any possibility of modelling this type of problem?

In the real world, this might translate to a shipment of hundreds of items that all need to be transported from A to B, but the total quantity exceeds any vehicle capacity, and there's no restriction on the size of any shipment, so having multiple small and multiple large deliveries is perfectly acceptable, as well as all large or all small deliveries (i.e. it's not possible to simply split the larger job up into naive smaller jobs of a certain size).

**Comment by reinterpretcat** [comment]
At the moment, there is no such built-in feature.

However, I think it can be modeled by splitting such shipment into smaller shipments and multiple vehicles should be able to pick multiple shipments at once (based on their available capacity). Post-processing can merge such splits into single job for each vehicle.

Theoretically, such approach can be even modeled natively on the solver side: preprocessing logic splits such jobs (customer should be able mark such jobs and preferable smallest demand). Post processing merges such activities into one for each vehicle. At least one thing to think about: do we need a dedicated objective, for example, to minimize amount of vehicles serving a "fat" job?

**Reply by Plastonick** [reply]
Thanks for the quick reply! Yes indeed, splitting the jobs into smaller jobs where each job's demand is the highest common factor of the vehicle capacities should do it, I might see if I can get that to work.

For my very particular use case I might find that problematic, for my use case vehicles should be fully loaded and then fully unloaded before re-loading again, I'm dealing with uncountable raw materials such as stone so loads cannot be mixed, which might happen with lots of small jobs. 

I've been modelling this currently as a jobs with a pickup then delivery task both with demands equal to the vehicle capacity to avoid mixing, but the problem has become more complex with multiple different possible vehicle capacities!

**Reply by reinterpretcat** [reply]
There is a `compatibility` feature which allows to setup a special property on the job which prevents jobs with different `compatibility` values to be assigned to the same tour.  Maybe it can help?

> compatibility (optional): compatibility class. Jobs with different compatibility classes cannot be assigned in the same tour. This is useful to avoid mixing cargo, such as hazardous goods and food.

https://reinterpretcat.github.io/vrp/concepts/pragmatic/problem/jobs.html

Please note, if you use `reload` feature, compatibility is applied on the whole tour, including all sub-trips (just missing implementation for reloads, can be extended)


Alternatively, you can try to play with job/vehicle skills: they support rules `one_of`, `all_of` and `none_of`

**Reply by Plastonick** [reply]
I did spot the `compatibility` functionality, but I think for my case I'd have to give each "large" job its own compatibility class, and then I'd thought that would prevent vehicles from working on multiple different jobs for the output solution, are you suggesting that's only the case if I'm using `reload`s but _not_ if I'm modelling my jobs as having both a `pickup` and `delivery` task? I'll have to have a play with that! 

Skills were something else I'd considered to handle this, and I think they will be useful for different concepts in my problem, but I'm not sure how I could use skills to keep multiple different types of material being loaded simultaneously.

**Reply by reinterpretcat** [reply]
Different jobs can share the same compatibility class, so it is not unique per unique job. 

Regarding reload, I meant that the compatibility works for the whole tour, not for sub-trips created by reload. So, if you don't use reload, then it doesn't matter.

Also, compatibility is specified for the whole job, so both pickup and delivery will get it.
